Python/specialize.c

```diff
@@ -0,0 +1,134 @@
+ * The first (index 0) cache entry is reserved for the count, to enable finding
+ * the first instruction from the base pointer.
+ * We use the SpecializedCacheOrInstruction union to refer to the data
+ * to avoid type punning.
+ */
+
+static SpecializedCacheOrInstruction *
+allocate(int cache_count, int instruction_count)
+{
+    assert(sizeof(SpecializedCacheOrInstruction) == 2*sizeof(int32_t));
+    assert(sizeof(SpecializedCacheEntry) == 2*sizeof(int32_t));
+    assert(cache_count > 0);
+    int count = cache_count + (instruction_count + INSTRUCTIONS_PER_ENTRY -1)/INSTRUCTIONS_PER_ENTRY;
+    SpecializedCacheOrInstruction *array = (SpecializedCacheOrInstruction *)
+        PyMem_Malloc(sizeof(SpecializedCacheOrInstruction) * count);
+    if (array == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+    array[0].entry.zero.cache_count = cache_count;
+    return array;
+}
+
+static int
+get_cache_count(SpecializedCacheOrInstruction *quickened) {
+    return quickened[0].entry.zero.cache_count;
+}
+
+static uint8_t adaptive[256] = { 0 };
+
+static uint8_t cache_requirements[256] = { 0 };
+
+static int
+entries_needed(_Py_CODEUNIT *code, int len)
+{
+    int cache_offset = 0;
+    for (int i = 0; i < len; i++) {
+        uint8_t opcode = _Py_OPCODE(code[i]);
+        uint8_t need = cache_requirements[opcode];
+        if (need == 0) {
+            continue;
+        }
+        assert(adaptive[opcode] != 0);
+        int oparg = oparg_from_offset_and_index(cache_offset, i);
+        assert(cache_offset == offset_from_oparg_and_index(oparg, i));
+        if (oparg < 0) {
+            cache_offset = i/2;
+        }
+        else if (oparg > 255) {
+            /* Cannot access required cache_offset */
+            continue;
+        }
+        cache_offset += need;
+    }
+    return cache_offset+1;
+}
+
+
+static inline _Py_CODEUNIT *
+first_instruction(SpecializedCacheOrInstruction *quickened)
+{
+    return &quickened[get_cache_count(quickened)].code[0];
+}
+
+static void
+optimize(SpecializedCacheOrInstruction *quickened, int len)
+{
+    _Py_CODEUNIT *instructions = first_instruction(quickened);
+    int cache_offset = 0;
+    int previous_opcode = -1;
+    for(int i = 0; i < len; i++) {
+        int opcode = _Py_OPCODE(instructions[i]);
+        uint8_t adaptive_opcode = adaptive[opcode];
+        if (adaptive_opcode && previous_opcode != EXTENDED_ARG) {
+            int oparg = oparg_from_offset_and_index(cache_offset, i);
+            if (oparg < 0) {
+                cache_offset = i/2;
+                oparg = 0;
+            }
+            if (oparg < 256) {
+                instructions[i] = _Py_CODEUNIT(adaptive_opcode, oparg);
+                cache_offset += cache_requirements[opcode];
+            }
+        }
+        else {
+            switch (opcode) {
+                /* Insert superinstructions here
+                 E.g.
+                case LOAD_FAST:
+                    if (previous_opcode == LOAD_FAST)
```

Comment:
Careful, what if the previous opcode was replaced by an adaptive one?

(Also maybe note that superinstructions don't reference the cache.)